File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& price) {
        int n = price.size();
        vector<vector<vector<int>>> dp(n + 1, vector<vector<int>>(3, vector<int>(2, 0)));

        for (int day = n - 1; day >= 0; day--) {
            for (int num_transactions = 0; num_transactions <= 2; num_transactions++) {
                for (int canBuy = 0; canBuy <= 1; canBuy++) {
                    int& profit = dp[day][num_transactions][canBuy];

                    int doNothing = dp[day + 1][num_transactions][canBuy];

                    if (canBuy) {
                        profit = max(doNothing, -price[day] + dp[day + 1][num_transactions][0]);
                    } else if (num_transactions < 2) {
                        profit = max(doNothing, price[day] + dp[day + 1][num_transactions + 1][1]);
                    }
                }
            }
        }

        return dp[0][0][1];
    }
};

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
    private int findMaxProfit(int day, int num_transactions, int canBuy, int[] price, int dp[][][]) {
        if(day == price.length || num_transactions == 2) return 0;

        if(dp[day][num_transactions][canBuy] != -1) return dp[day][num_transactions][canBuy];

        int profit = 0;

        int doNothing = findMaxProfit(day + 1, num_transactions, canBuy, price, dp);

        if(canBuy == 1) {
            profit += Math.max(doNothing, -price[day] + findMaxProfit(day + 1, num_transactions, 0, price, dp));
        }

        else profit += Math.max(doNothing, price[day] + findMaxProfit(day + 1, num_transactions + 1, 1, price, dp));

        return dp[day][num_transactions][canBuy] = profit;
    }
    
    public int maxProfit(int[] price) {
        int n = price.length;
        
        int dp[][][] = new int[n + 1][3][2];
        for(int i = 0; i < n + 1; i++) {
            for(int j = 0; j < 3; j++) {
                for(int k = 0; k < 2; k++) {
                    dp[i][j][k] = -1;
                }
            }
        }
        
        return findMaxProfit(0, 0, 1, price, dp);
    }
}